app/validators/digital_object/enabled_dynamic_fields_validations.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module DigitalObject::EnabledDynamicFieldsValidations
+  # Returns any errors related to project level validations.
+  #
+  # @return [String] if there are errors
+  # @return false if there are no errors
+  def enabled_field_errors(digital_object, data)
+    errors = []
+    dynamic_field_paths = []
+    data.each do |df_group, children|
+      dynamic_field_paths = collect_field_paths(df_group, children, dynamic_field_paths)
+      dynamic_field_paths.each do |dynamic_field_path|
+        if (e = not_enabled?(dynamic_field_path, digital_object))
+          errors.concat e.map { |i| [dynamic_field_path, i] }
+        end
+      end
+    end
+
+    errors = check_required_fields(dynamic_field_paths, digital_object, errors)
+
+    errors
+  end
+
+  def not_enabled?(dynamic_field_path, digital_object)
+    no_match = true
+    dynamic_field = DynamicField.find_by(path: dynamic_field_path)
+    if dynamic_field && EnabledDynamicField.where(dynamic_field: dynamic_field.id,
+                                project: digital_object.primary_project,
+                                digital_object_type: digital_object.digital_object_type).first
+      no_match = false
+    end
+    return ['field must be enabled'] if no_match
+    false
+  end
+
+  def check_required_fields(dynamic_field_paths, digital_object, errors)
+    required_enabled_fields = EnabledDynamicField.where(project: digital_object.primary_project,
```

Comment:
```suggestion
    required_fields = DynamicField.where(
      id: EnabledDynamicField.where(
        project: digital_object.primary_project, digital_object_type: digital_object.digital_object_type, required: true
      ).pluck(:dynamic_field_id)
    )
```
Since you're only using the underling dynamic_field property for each of the EnabledDynamicField objects, you can boost efficiency here by retrieving all of the relevant fields in one go rather than calling the `.dynamic_field` method on each (which I believe will lazy load the field, resulting in an extra SQL query for each iteration).